Conflicting suggestions for unit_arg lint between clippy and (nightly) rustc

Given the following (contrived) code:
```rust
#![allow(unused)]
use std::io;

fn unit_fn() -> io::Result<()> {
    Ok(())
}

fn fallible() -> io::Result<()> {
    Ok(unit_fn()?)
}

fn main() {}
```
[Playground](https://play.rust-lang.org/?gist=e2dbb849ddbf8d83857fbd55edc00b8e&version=nightly&mode=debug&edition=2015)

clippy lints this as:
```
   Checking playground v0.0.1 (file:///playground)
warning: passing a unit value to a function
 --> src/main.rs:9:8
  |
9 |     Ok(unit_fn()?)
  |        ^^^^^^^^^^
  |
  = note: #[warn(unit_arg)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#unit_arg
help: if you intended to pass a unit value, use a unit literal instead
  |
9 |     Ok(())
  |        ^^

    Finished dev [unoptimized + debuginfo] target(s) in 1.35s
```
If I intentionally change the code to the following (incorrect) version:
```rust
#![allow(unused)]
use std::io;

fn unit_fn() -> io::Result<()> {
    Ok(())
}

fn fallible() -> io::Result<()> {
    unit_fn()?
}

fn main() {}
```
I get a compiler error (on current nightly):
```
   Compiling playground v0.0.1 (file:///playground)
error[E0308]: try expression alternatives have incompatible types
 --> src/main.rs:9:5
  |
9 |     unit_fn()?
  |     ^^^^^^^^^^
  |     |
  |     expected enum `std::result::Result`, found ()
  |     help: try wrapping with a success variant: `Ok(unit_fn()?)`
  |
  = note: expected type `std::result::Result<(), std::io::Error>`
             found type `()`
```
So it seems like rustc is suggesting a form that clippy lints against, i.e.:
```
help: try wrapping with a success variant: `Ok(unit_fn()?)`
```
Note that the `stable` compiler does not have that `help: try wrapping ...` suggestion, so it's something recently added.

Also, to be clear - I know the code is incorrect and can be written differently.  My only point is to illustrate the discrepancy between syntactic/stylistic forms that clippy and rustc suggest.

--- a/clippy_lints/src/types.rs
+++ b/clippy_lints/src/types.rs
@@ -609,36 +609,43 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnitArg {
         if in_macro(expr.span) {
             return;
         }
+
+        // apparently stuff in the desugaring of `?` can trigger this
+        // so check for that here
+        // only the calls to `Try::from_error` is marked as desugared,
+        // so we need to check both the current Expr and its parent.
+        if is_questionmark_desugar_marked_call(expr) {
+            return;
+        }
+        if_chain! {
+            let map = &cx.tcx.hir();
+            let opt_parent_node = map.find(map.get_parent_node(expr.id));
+            if let Some(hir::Node::Expr(parent_expr)) = opt_parent_node;
+            if is_questionmark_desugar_marked_call(parent_expr);
+            then {
+                return;
+            }
+        }
+
         match expr.node {
             ExprKind::Call(_, ref args) | ExprKind::MethodCall(_, _, ref args) => {
                 for arg in args {
                     if is_unit(cx.tables.expr_ty(arg)) && !is_unit_literal(arg) {
-                        let map = &cx.tcx.hir();
-                        // apparently stuff in the desugaring of `?` can trigger this
-                        // so check for that here
-                        // only the calls to `Try::from_error` is marked as desugared,
-                        // so we need to check both the current Expr and its parent.
-                        if !is_questionmark_desugar_marked_call(expr) {
-                            if_chain! {
-                                let opt_parent_node = map.find(map.get_parent_node(expr.id));
-                                if let Some(hir::Node::Expr(parent_expr)) = opt_parent_node;
-                                if is_questionmark_desugar_marked_call(parent_expr);
-                                then {}
-                                else {
-                                    // `expr` and `parent_expr` where _both_ not from
-                                    // desugaring `?`, so lint
-                                    span_lint_and_sugg(
-                                        cx,
-                                        UNIT_ARG,
-                                        arg.span,
-                                        "passing a unit value to a function",
-                                        "if you intended to pass a unit value, use a unit literal instead",
-                                        "()".to_string(),
-                                        Applicability::MachineApplicable,
-                                    );
-                                }
+                        if let ExprKind::Match(.., match_source) = &arg.node {
+                            if *match_source == MatchSource::TryDesugar {
+                                continue;
                             }
                         }
+
+                        span_lint_and_sugg(
+                            cx,
+                            UNIT_ARG,
+                            arg.span,
+                            "passing a unit value to a function",
+                            "if you intended to pass a unit value, use a unit literal instead",
+                            "()".to_string(),
+                            Applicability::MachineApplicable,
+                        );
                     }
                 }
             },

--- a/tests/ui/unit_arg.fixed
+++ b/tests/ui/unit_arg.fixed
@@ -47,6 +47,17 @@ fn question_mark() -> Result<(), ()> {
     Ok(())
 }
 
+#[allow(dead_code)]
+mod issue_2945 {
+    fn unit_fn() -> Result<(), i32> {
+        Ok(())
+    }
+
+    fn fallible() -> Result<(), i32> {
+        Ok(unit_fn()?)
+    }
+}
+
 fn main() {
     bad();
     ok();

--- a/tests/ui/unit_arg.rs
+++ b/tests/ui/unit_arg.rs
@@ -54,6 +54,17 @@ fn question_mark() -> Result<(), ()> {
     Ok(())
 }
 
+#[allow(dead_code)]
+mod issue_2945 {
+    fn unit_fn() -> Result<(), i32> {
+        Ok(())
+    }
+
+    fn fallible() -> Result<(), i32> {
+        Ok(unit_fn()?)
+    }
+}
+
 fn main() {
     bad();
     ok();

